File: JavaScript/huffmanEncoder.js
const Node = require('./Node').Node;

function frequencyMap(text) {
	let obj = {};
	for (let char of text) {
		obj[char] = obj[char] + 1 || 1
	}
	return obj;
}

function mapAndSort(text) {
	let map = frequencyMap(text);
	const sortable = [];
	for(let obj in map) {
		sortable.push([obj, map[obj]]);
	}
	sortable.sort((x,y) => x[1] - y[1]);
	const nodes = sortable.map((x) => new Node(x))
	return nodes;
}

function minHeap(nodes) {
	while(nodes.length >= 2) {
		let one = nodes.shift();
		let two = nodes.shift();
		let sum = one.data[1] + two.data[1];
		if (one.data[1] <= two.data[1]) {
			nodes.push(new Node(['',sum], one, two));
		} else {
			nodes.push(new Node(['',sum], two, one));
		}
		nodes.sort((x,y) => x.data[1] - y.data[1]);
	}
	return nodes[0];
}

function mapPrefixes(map, root, path='') {
	if (!root.left && !root.right) {
		map[root.data[0]] = path;
	} else {
		mapPrefixes(map, root.left, path+'0');
		mapPrefixes(map, root.right, path+'1');
	}
}

function generatePrefixes(text, tree) {
	const prefixes = {}
	mapPrefixes(prefixes,tree);
	return prefixes;
}

function compress(string) {
	const size = 8;
	const byteArray = new Uint8Array(Math.ceil(string.length/8));
	let index = 0;
	let byte = 0;
	let holder = '00000000'
	while(index < string.length) {
		let slice = string.slice(index, index + size);
		if(slice.length < 8) {
			slice = slice + holder.substr(slice.length);
		}
		byteArray[byte] = parseInt(slice, 2);
		index+=size;
		byte++;
	}
	return byteArray;
}


function encodeString(text) {
	let result = ''
	const tree = minHeap(mapAndSort(text + '■'));
	const prefixes = generatePrefixes(text, tree);
	console.log(prefixes);
	const freqMap = frequencyMap(text + '■');
	for(let char of text) {
		for(let code in prefixes) {
			if (char === code) {
				result += prefixes[code];
			}
		}
	}
	const byteArray = compress(result);
	return {huffmanString: result, key: tree, byteArray: byteArray, header: freqMap};
}

module.exports = encodeString;

File: Python/HuffmanEncoder.py
from node import Node

def frequencyMap(text):
	map = {}
	for char in text:
		if char in map:
			map[char] += 1
		else:
			map[char] = 1
	return map

def sortAndCreateNodes(freqMap):
	sortable = []
	for obj in freqMap:
		sortable.append([obj, freqMap[obj]])
	sortable.sort(key=lambda x: x[1])
	nodes = list(map(Node, sortable))
	return nodes

def minHeap(nodes):
	while len(nodes) >= 2:
		one = nodes.pop(0)
		two = nodes.pop(0)
		total = one.data[1] + two.data[1]
		if one.data[1] <= two.data[1]:
			nodes.append(Node(['',total], one, two))
		else:
			nodes.append(Node(['',total], two, one))
		nodes.sort(key=lambda x: x.data[1])
	return nodes[0]

def mapPrefixes(map, root, path=''):
	if root.left is None and root.right is None:
		map[root.data[0]] = path
	else:
		mapPrefixes(map, root.left, path+'0')
		mapPrefixes(map, root.right, path+'1')

def generatePrefixes(tree):
	prefixes = {}
	mapPrefixes(prefixes, tree)
	return prefixes

def createByteArray(text):
	bArr = bytearray()
	holder = '00000000'
	for x in range(0, len(text), 8):
		chunk = text[x:x+8]
		if len(chunk) < 8:
			chunk = chunk + holder[len(chunk):]
		bArr.append(int(chunk, 2))
	return bArr


def writeFileHeader(freqMap):
	result = ''
	for char in freqMap:
		result += (char + str(freqMap[char]) + '_')
	result = result[:len(result) - 1]
	return result
	
def encodeString(text):
	result = ''
	data = text + '■'
	freqMap = frequencyMap(data)
	tree = minHeap(sortAndCreateNodes(freqMap))
	prefixes = generatePrefixes(tree)
	for char in data:
		result += prefixes[char]
	byteArr = createByteArray(result)
	header = writeFileHeader(freqMap)
	return {'string': result, 'tree': tree, 'byteArr': byteArr, 'header': header}




